addons/stock/models/stock_move_line.py

```diff
@@ -154,6 +154,16 @@ def _get_similar_move_lines(self):
             lines |= picking_id.move_line_ids.filtered(lambda ml: ml.product_id == self.product_id and (ml.lot_id or ml.lot_name))
         return lines
 
+    @api.model_cr
+    def init(self):
+        if not tools.index_exists(self._cr, 'stock_move_line_free_reservation_index'):
+            self._cr.execute("""
+                CREATE INDEX stock_move_line_free_reservation_index
+                ON
+                    stock_move_line (id, product_id, lot_id, location_id, owner_id, package_id)
+                WHERE
+                    (state IS NULL OR state NOT IN ('cancel', 'done')) AND product_qty > 0""")
```

Comment:
Isn't `(state IS NULL OR state NOT IN ('cancel', 'done'))` redundant? `state NOT IN ('cancel', 'done')` implies that it can be `NULL`, right?

Reply:
you always need to test NULL directly, there is some magic operator (`

```psql
13e=# SELECT 1 WHERE NULL NOT IN ('cancel');
 ?column? 
----------
(0 rows)

13e=# SELECT 1 WHERE NULL IN ('cancel');
 ?column? 
----------
(0 rows)
```

(ORM does it for us most of the time)